### PR TITLE
Add support for documenting values of constants.

### PR DESCRIPTION
**Pull request by jpeach**
A common pattern in CRD APIs is to use enum validation and a
corresponding Go type that has constant values declared for
each enumerated value.

This change detects constants that are declared as instances of
an underlying type, and allows the type template to format the
values and comments in to a table.

There are 3 parts to the change:

1. Capture the constant types in a new field on the `apiPackage`.
2. Export the constant value through the `typeDisplayName` helper.
3. Export the constants of a given type through `constantsOfType`.

Signed-off-by: James Peach <jpeach@apache.org>

### CODE REVIEW

**Comment by ahmetb** [comment]
Sounds great. If you have any example output, can you upload to a gist for rendering?

**Comment by jpeach** [comment]
Here's an example rendered from this service-apis docs:

Screen Shot 2020-12-14 at 6 01 39 pm

File: main.go [hunk 1]
```diff
@@ -433,6 +447,20 @@ func tryDereference(t *types.Type) *types.Type {
 	return t
 }
 
+// finalUnderlyingTypeOf walks the type hierarchy for t and returns
+// its base type (i.e. the type that has no further underlying type).
+func finalUnderlyingTypeOf(t *types.Type) *types.Type {
+	u := t
```
Comment (ahmetb) [inline main.go:453]: variable seems superfluous but ok

File: template/type.tpl [hunk 2]
```diff
@@ -22,6 +22,26 @@
...
+    <thead>
+        <tr>
+            <th>Value</th>
+            <th>Description</th>
+        </tr>
+    </thead>
+    <tbody>
+      {{- range . -}}
+      <tr>
+        {{- /* renderComments implicitly creates a <p> element, so we do the same here to make the value line up nicely.  */ -}}
```
Comment (ahmetb) [inline template/type.tpl:36]: I don't get what this comment says, can you clarify the problem again please?

**Review by ahmetb — APPROVED** [review]

File: template/type.tpl (hunk 2 above)
Comment (jpeach) [reply template/type.tpl:36]: `renderComments` wraps the comments in a `<p>` element via the markdown rendered. In a table row, if one cell is wrapped in a `<p>` but the other isn't they misalign and it looks bad.

File: main.go (hunk 1 above)
Comment (jpeach) [reply main.go:453]: Fixed.

File: template/type.tpl (hunk 2 above)
Comment (jpeach) [reply template/type.tpl]: Hopefully this is clearer now.

**Comment by RichieEscarez** [comment]
@ahmetb 

FYI: In Knative ([v0.20.0](https://github.com/knative/serving/releases/tag/v0.20.0)), picking up this change fixed the following error for us:

```
F0112 17:03:31.667704   19198 main.go:160] failed: template: type.tpl:24: function "constantsOfType" not defined
parse error
main.render
	/Users/rescarez/go/pkg/mod/github.com/ahmetb/gen-crd-api-reference-docs@v0.2.0/main.go:619
main.main.func1
	/Users/rescarez/go/pkg/mod/github.com/ahmetb/gen-crd-api-reference-docs@v0.2.0/main.go:143
main.main
	/Users/rescarez/go/pkg/mod/github.com/ahmetb/gen-crd-api-reference-docs@v0.2.0/main.go:158
runtime.main
	/usr/local/Cellar/go/1.15.3/libexec/src/runtime/proc.go:204
runtime.goexit
	/usr/local/Cellar/go/1.15.3/libexec/src/runtime/asm_amd64.s:1374
failed to render the result
main.main.func1
	/Users/rescarez/go/pkg/mod/github.com/ahmetb/gen-crd-api-reference-docs@v0.2.0/main.go:145
main.main
	/Users/rescarez/go/pkg/mod/github.com/ahmetb/gen-crd-api-reference-docs@v0.2.0/main.go:158
runtime.main
	/usr/local/Cellar/go/1.15.3/libexec/src/runtime/proc.go:204
runtime.goexit
	/usr/local/Cellar/go/1.15.3/libexec/src/runtime/asm_amd64.s:1374

```

**Comment by ahmetb** [comment]
@RichieEscarez ack we probably need to lock down the code and the template in Knative side so we don't get impacted by such updates

**Comment by ahmetb** [comment]
I've just made a new release to the tool v0.3.0. Try it again please.
There might still be an issue with how we lock down versions in Knative repo. The best course of action would be to lock down both the "template" file and the tool binary to a particular version, so it doesn't keep moving as we release new versions here.

**Pull request merged**